lib/api/apiUtils/bucket/bucketCors.js

```diff
@@ -0,0 +1,278 @@
+        }
+        if (id[0] === '') {
+            return undefined;
+        }
+        // NOTE: AWS does not make this check, but the spec specifies IDs should
+        // be up to 255 characters in length.
+        // http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTcors.html
+        if (id[0].length > 255) {
+            return errors.InvalidRequest.customizeDescription(errMsg);
+        }
+        return true; // to indicate ID exists
+    },
+    validateMaxAgeSeconds(seconds) {
+        if (!seconds) {
+            return undefined;
+        }
+        if (!Array.isArray(seconds) || seconds.length !== 1) {
+            return errors.MalformedXML;
+        }
+        if (seconds[0] === '') {
+            return undefined;
+        }
+        const parsedValue = parseInt(seconds[0], 10);
+        const errMsg = `MaxAgeSeconds "${seconds[0]}" is not a valid value.`;
+        if (isNaN(parsedValue) || parsedValue < 0) {
+            return errors.MalformedXML.customizeDescription(errMsg);
+        }
+        return parsedValue;
+    },
+    validateNumberRules(length) {
+        if (length > 100) {
+            return errors.InvalidRequest.customizeDescription(errs.numberRules);
+        }
+        return undefined;
+    },
+    validateOriginAndMethodExist(allowedMethods, allowedOrigins) {
+        if (allowedOrigins && allowedMethods &&
+        Array.isArray(allowedOrigins) &&
+        Array.isArray(allowedMethods) &&
+        allowedOrigins.length > 0 &&
+        allowedMethods.length > 0) {
+            return undefined;
+        }
+        return errors.MalformedXML
+            .customizeDescription(errs.originAndMethodExist);
+    },
+    validateMethods(methods) {
+        let invalidMethod;
+        function isValidMethod(method) {
+            if (method !== 'GET' && method !== 'PUT' && method !== 'HEAD' &&
+            method !== 'POST' && method !== 'DELETE') {
+                invalidMethod = method;
+                return false;
+            }
+            return true;
+        }
+        if (!methods.every(isValidMethod)) {
+            const errMsg = 'Found unsupported HTTP method in CORS config. ' +
+            `Unsupported method is "${invalidMethod}"`;
+            return errors.InvalidRequest.customizeDescription(errMsg);
+        }
+        return undefined;
+    },
+    validateOrigins(origins) {
+        for (let i = 0; i < origins.length; i++) {
+            const origin = origins[i];
+            const errMsg = `AllowedOrigin "${origin}" can not have ` +
+            'more than one wildcard.';
+            if (typeof origin !== 'string' || origin === '') {
+                return errors.MalformedXML;
+            }
+            const numberWildcards = (origin.match(/\*/g) || []).length;
+            if (numberWildcards > 1) {
+                return errors.InvalidRequest.customizeDescription(errMsg);
+            }
+        }
+        return undefined;
+    },
+    validateAllowedHeaders(headers) {
```

Comment:
Do you think we should check for invalid symbols (anything non-alphanumeric or a dash) in `AllowedHeader`s? AWS performs this check for `ExposeHeader`s (the message they return is actually `AllowedHeader [value] contains invalid character.` which makes me suspect they intended to check). That's the only reason why I might not convert validateOrigins into a utility check I can also call in `validateAllowedHeaders`, because I'd have to iterate through the list of AllowedHeaders a second time.

On the other hand, the only way I can think to perform that check is to use a regex test (like I did for `ExposeHeader`s) and since AWS doesn't actually perform the check, maybe we don't need to.